server/utils/cloudFunctions/functions/test/algoliaTests/algoliaIntegrationTests.test.js

```diff
@@ -0,0 +1,218 @@
+ *       and a string equality match is alternatively much easier. 
+ *     - For the above reason, `getObjects` is used with singleton lists
+ *       to verify that a record has been deleted instead of `getObject`. 
+ * - Since error handling is managed by Firestore instead of the cloud 
+ *   functions, it seems reasonable to not test for invalid inputs or 
+ *   other failures (per Aileme.)
+ * - `setTimeout` is used to allow firebase changes to propogate to algolia. 
+ *   - `setTimeout` requires that `done()` be called at the end of the 
+ *     inner function body. It is not equivalent to `return`. 
+ * - `TIMEOUT_10S` was chosen because this is the maximum amount of latency 
+ *   for firebase to execute a function. 
+ * - `TIMEOUT_15S` is used to extend the default 2000ms limit set by mocha.
+ * - The overall test timeout limit must be longer than the `setTimeout`
+ *   method to allow time for the algolia api call.    
+ * Tests are structured with promise nesting. 
+ * - There is a debate about nesting vs chaining promises. In this case, 
+ *     nesting is advantageous because it easily keeps everything in scope
+ *     without much fuss. Thus, nesting is fine to use. 
+ * - Also this is how the official docs do it. 
+ * Author: Robert Marcus
+ * Date: July 16, 2020
+ */
+
+
+// Import the testing environment configuration.
+const { chai,
+        TIMEOUT_10S,
+        TIMEOUT_15S,
+        DEV_ADS_COLLECTION } = require('../testConfig');
+
+const { deleteCollection } = require('../deleteCollection');
+
+// Import firebase environment configuration.
+const { DB } = require('../../firebaseConfig');
+
+// Import algolia dev environment configuration.
+const { DEV_ADS_INDEX } = 
+    require('../../algoliaCloudFunctions/algoliaConfig');
+
+describe("Algolia integrations tests", () => {
+  // After every 'describe' block, reset the test environments. 
+  // Since these are live dev db's, this process can be very flaky.
+  // In general, don't expect an 'advertiser' will be reset from 
+  // test to test. 
+  after(() => {
+    deleteCollection(DB, 'dev_ads');
+  });
+
+  describe("test_createRecord", () => {
+    it("should propogate an ad document into algolia index", function(done) {
+      const ad = {advertiser: "a", startDate: "2019-10-15"};
+
+      DEV_ADS_COLLECTION.doc("a").set(ad).then(() => {
+        setTimeout(function(){
+          DEV_ADS_INDEX.getObject("a").then(content => {
+            chai.expect(content.data).to.deep.equal(ad);
+          }).catch(err => console.log(err));
+
+          done();
+        }, TIMEOUT_10S);
+      });
+    }).timeout(TIMEOUT_15S);
+    
+    it("should propogate multiple ad documents without overwriting", function(done) {
+      const adList = [{advertiser: "b", startDate: "2019-10-15"},
+                      {advertiser: "c", startDate: "2019-10-15"},
+                      {advertiser: "d", startDate: "2019-10-16"}];
+
+      DEV_ADS_COLLECTION.doc("b").set(adList[0]).then(() => {
```

Comment:
Having callback hell like this is not ideal at all, is there other way we can get away with this (i.e. using Await?)

Reply:
I agree, this is a rats nest of callback nonsense. I have experimented with using  await and such, and I've also experimented with using stuff like `Promise.all`. 
- The main problem I had with using `await` was significantly more test flakiness. 
- I had weird issues using `Promise.all` with `chai` where assertions wouldn't fail properly. 

EOD, promise nesting makes two things very easy: keeps things in scope easily, makes sure things get executed in the proper order only after completion. It's also a lot less flaky than using await. Since this is just for integration testing and not code that will need to change much ever, I think it's an acceptable compromise for the ugliness.  

Going to leave this comment `unresolved` for visibility in case others have a similar q.